numba/targets/arrayobj.py

```diff
@@ -3765,7 +3765,23 @@ def _arange_dtype(*args):
     elif any(isinstance(a, types.Float) for a in bounds):
         dtype = types.float64
     else:
-        dtype = types.intp
+        # It's not possible for the dtype to replicate NumPy as integer literals
+        # are intp size in Numba but np.dtype(int) (i.e. intc) in NumPy. The
+        # NumPy logic is to basically do max(np.long, *[type(x) for x in args]).
+        # On windows 64, in NumPy, that would mean:
+        # * np.arange(1, 10) -> max(np.long, np.intc, np.intc) -> int32
+        # but in Numba:
+        # * np.arange(1, 10) -> max(np.long, np.intp, np.intp) -> int64
+        #
+        # It's therefore not possible to replicate the case where integer
+        # literals are supplied, however best effort is made to correctly handle
+        # cases like:
+        # * np.arange(np.int8(10)) -> max(np.long, np.int8) -> np.long
+        #
+        # Alg ref:
+        # https://github.com/numpy/numpy/blob/maintenance/1.17.x/numpy/core/src/multiarray/ctors.c#L3376-L3377    # noqa: E501
+        dtype = max(bounds + [types.long_,])
+        print(bounds, "resolved", dtype, "long = ", types.long_)
```

Comment:
Leftover print statement to remove here.

Reply:
tests are failing, I put that in to see what was going on :)

Reply:
I thought it was this: https://github.com/numba/numba/pull/4882/files#diff-81d8f85ead1fb3ae948c94ed8d2a969eR3768-R3782 but it seems like a `types.long_` is appearing as an `int64` on linux32 and NumPy wants an `int32`.
https://dev.azure.com/numba/numba/_build/results?buildId=3317&view=logs&jobId=536d9229-f68f-5476-bdce-7191d78ec342&taskId=695d2d88-17fa-5407-392b-a8e99ffa20e2&lineStart=317&lineEnd=321&colStart=1&colEnd=1

Reply:
ah, makes sense